File: android/src/main/java/com/reactnativecouchbaselite/RNReplicator.kt
package com.reactnativecouchbaselite

import arrow.core.*
import arrow.core.extensions.either.applicative.applicative
import com.couchbase.lite.*
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.WritableMap

object RNReplicator {
  private val replicators: HashMap<String, Replicator> = hashMapOf()
  private val listenerTokens: HashMap<String, ListenerToken> = hashMapOf()

  private fun getOne(database: String) = replicators[database]
    .rightIfNotNull { "replicator for $database not found" }

  private fun writeStatus(status: AbstractReplicator.Status): WritableMap {
    val result = Arguments.createMap()
    result.putString("activityLevel", status.activityLevel.name)
    result.putInt("completed", status.progress.completed.toInt())
    result.putInt("total", status.progress.total.toInt())
    status.error.toOption().fold(
      { result.putNull("error") },
      { error ->
        val errorMap = Arguments.createMap()
        errorMap.putInt("code", error.code)
        errorMap.putString("domain", error.domain)
        errorMap.putString("message", error.localizedMessage)
        result.putMap("error", errorMap)
      }
    )
    return result
  }

  private fun writeStatusEvent(eventId: String, status: AbstractReplicator.Status): WritableMap {
    val result = Arguments.createMap()
    result.putString("id", eventId)
    result.putMap("status", writeStatus(status))
    return result
  }

  fun run(payload: SafeReadableMap, eventEmitter: EventEmitter): Either<String, WritableMap> =
    payload.tag.flatMap { tag ->
      fun getReplicator() = payload.getString("database")
        .flatMap { getOne(it) }
      when (tag) {
        "debug" -> {
          Database.setLogLevel(LogDomain.REPLICATOR, LogLevel.VERBOSE)
          Either.right(Arguments.createMap())
        }
        "init" -> payload.getMap("config")
          .flatMap { RNReplicatorConfiguration.decode(it) }
          .flatMap { config ->
            replicators[config.database.name]
              .rightIfNull { "$config.database.name already init" }
              .map {
                val replicator = Replicator(config)
                replicators[config.database.name] = replicator
                Arguments.createMap()
              }
          }

        "start" -> getReplicator()
          .map { it.start() }
          .map { Arguments.createMap() }

        "stop" -> getReplicator()
          .map { it.stop() }
          .map { Arguments.createMap() }

        "status" -> getReplicator()
          .map { writeStatus(it.status) }

        "addChangeListener" -> Either.applicative<String>().tupled(
          getReplicator(),
          payload.getString("eventId")
        ).fix()
          .map { tuple ->
            val replicator = tuple.a
            val eventId = tuple.b
            listenerTokens[eventId] = replicator.addChangeListener { change ->
              eventEmitter.sendEvent("Replicator.Status", writeStatusEvent(eventId, change.status))
            }
            Arguments.createMap()
          }

        "removeChangeListener" -> Either.applicative<String>().tupled(
          getReplicator(),
          payload.getString("eventId")
        ).fix()
          .flatMap { tuple ->
            val replicator = tuple.a
            val eventId = tuple.b
            listenerTokens[eventId].rightIfNotNull {
              "listenerToken for eventId $eventId not found"
            }.map { token ->
              replicator.removeChangeListener(token)
              listenerTokens.remove(eventId)
              Arguments.createMap()
            }
          }

        else -> Either.left("unknown replicator tag $tag")
      }
    }
}

File: ios/RNReplicator.swift
import Bow
import CouchbaseLiteSwift

import Foundation

class RNReplicator {
  static var replicators: [String: Replicator] = [:]
  static var listenerTokens: [String: ListenerToken] = [:]
  
  static func getOne(_ database: String) -> Either<String, Replicator> {
    return replicators[database].rightIfNotNull("replicator for \(database) not found")
  }
  
  static func writeActivityLevel(_ activity: Replicator.ActivityLevel) -> String {
    switch activity {
    case .stopped:
      return "STOPPED"
    case .offline:
      return "OFFLINE"
    case .connecting:
      return "CONNECTING"
    case .idle:
      return "IDLE"
    case .busy:
      return "BUSY"
    }
  }
  
  static func writeStatus(_ status: Replicator.Status) -> [AnyHashable: Any] {
    var result = SafeDictionary.empty()
    
    result["activityLevel"] = writeActivityLevel(status.activity)
    result["completed"] = status.progress.completed
    result["total"] = status.progress.total
    
    (status.error as NSError?).toOption().fold(
      { result["error"] = nil },
      { error in
        var errorMap = SafeDictionary.empty()
        errorMap["code"] = error.code
        errorMap["domain"] = error.domain
        errorMap["message"] = error.localizedDescription
        result["error"] = errorMap
      }
    )
    return result
  }
  
  static func writeStatusEvent(_ eventId: String, _ status: Replicator.Status) -> [AnyHashable: Any] {
    var result = SafeDictionary.empty()
    result["id"] = eventId
    result["status"] = writeStatus(status)
    return result
  }
  
  static func run(_ payload: SafeDictionary) -> Either<String, [AnyHashable: Any]> {
    return payload.getString("tag")
      .flatMap({ (tag) -> Either<String, [AnyHashable: Any]> in
        func getReplicator() -> Either<String, Replicator> {
          return payload.getString("database")
            .flatMap(getOne)^
        }
        
        switch tag {
        case "debug":
          Database.setLogLevel(.verbose, domain: .replicator)
          return Either.right(SafeDictionary.empty())
          
        case "init":
          return payload.getDict("config")
            .flatMap(RNReplicatorConfiguration.decode)
            .flatMap({ (config: ReplicatorConfiguration) -> Either<String, [AnyHashable: Any]> in
              let database = config.database.name
              return replicators[database]
                .rightIfNotNull("")
                .swap()
                .mapLeft({ _ in "\(config.database.name) already init" })
                .map({ (_) -> [AnyHashable: Any] in
                  let replicator = Replicator(config: config)
                  replicators[database] = replicator
                  return SafeDictionary.empty()
                })^
            })^
          
        case "start":
          return getReplicator()
            .map({ $0.start() })
            .map({ SafeDictionary.empty() })^
          
        case "stop":
          return getReplicator()
            .map({ $0.stop() })
            .map({ SafeDictionary.empty() })^

        case "status":
          return getReplicator()
            .map({ $0.status })
            .map(writeStatus)^
          
        case "addChangeListener":
          return Either<String, [AnyHashable: Any]>.map(
            getReplicator(),
            payload.getString("eventId"),
            { (replicator, eventId) in
              listenerTokens[eventId] = replicator.addChangeListener { change in
                EventEmitter.sharedInstance.dispatch(name: "Replicator.Status", body: writeStatusEvent(eventId, change.status))
              }
              return SafeDictionary.empty()
            }
          )^
          
        case "removeChangeListener":
          return Either<String, [AnyHashable: Any]>.map(
            getReplicator(),
            payload.getString("eventId"),
            { (replicator, eventId) in
              listenerTokens[eventId].rightIfNotNull("listenerToken for eventId \(eventId) not found")
                .map({ token in
                  replicator.removeChangeListener(withToken: token)
                  listenerTokens.removeValue(forKey: eventId)
                  return SafeDictionary.empty()
                })
            }
            ).flatMap({ $0 })^
        default:
          return Either.left("unknown replicator tag \(tag)")
        }
      })^
  }
}
